### DISCUSSION TITLE
dotdirs (.Sent and so on) seem to be enforced for the maildir backend

### DISCUSSION THREAD

**Discussion by polyzen** [pimalaya/himalaya#327] [General]
Curious to know if there's a reason for this?

**Comment by soywod** [comment]
What do you mean by "enforced"? In the regular Maildir spec, a maildir folder contains 3 subfolders `tmp`, `cur` and `new`. The rest is skipped. Files starting by a dot (`.`) are also ignored. The [Maildir++](http://www.courier-mta.org/imap/README.maildirquota.html) introduces the fact to accept subfolders starting by a dot (`.`) as "mailboxes":

> Maildir++ adds a couple of things to a standard Maildir: folders and quotas. […] Folders are created as subdirectories under the main Maildir. The name of the subdirectory always starts with a period. For example, a folder named "Important" will be a subdirectory called ".Important". You can't have subdirectories that start with two periods.

**Reply by polyzen** [reply]
The maildir backend expects the mailboxes to start with a dot, eg. with `maildir-dir = "~/.mail/arch"`:
```
Error: cannot get maildir instance from "INBOX"

Caused by:
    cannot read maildir directory "/home/user/.mail/arch/.INBOX"
```

**Reply by soywod** [reply]
It should work with lowercase mailbox "inbox", try sth like `himalaya -m inbox list`. The bug is related to the case of the inbox folder: the IMAP backend uses by default `INBOX`, but it should be `inbox` for maildir. I opened an issue https://github.com/soywod/himalaya/issues/330.

**Reply by polyzen** [reply]
> It should work with lowercase mailbox "inbox", try sth like himalaya -m inbox list.

~~This works.~~

> The bug is related to the case of the inbox folder: the IMAP backend uses by default INBOX, but it should be inbox for maildir

mbsync uses `INBOX`: ["The special mailbox INBOX exists in every Store .."](https://isync.sourceforge.io/mbsync.html#CONFIGURATION).

**Reply by polyzen** [reply]
> > It should work with lowercase mailbox "inbox", try sth like himalaya -m inbox list.
> 
> This works.

Spoke too soon, had the IMAP backend still in my config ;D.

**Reply by polyzen** [reply]
maildir backend:
`himalaya -m inbox list` outputs just:
`HASH │FLAGS │SUBJECT │SENDER │DATE`
`himalaya -m INBOX list` leads to #329.

notmuch backend:
`himalaya -m inbox list` leads to #329.

**Reply by polyzen** [reply]
For the maildir backend, `himalaya list` would also lead to #329, but only if I renamed the `INBOX` dir to `.INBOX`.

**Reply by soywod** [reply]
Both issues on maildir and notmuch should be fixed, please let me know if it works well for you too!

**Reply by polyzen** [reply]
notmuch backend works great.

maildir backend doesn't find anything with https://github.com/soywod/himalaya/commit/d79c6c40a7e3e029cac979c6cd39cc67a5da15ba but does when built from https://github.com/soywod/himalaya/commit/984eb8c9f7c6374bb6279a7c9c3ae451182dc295:
- Config: https://paste.xinu.at/fd5Lg/
- Log: https://paste.xinu.at/0o3O/

The `cannot read maildir directory "/home/user/.mail/arch/.INBOX"` error still occurs, but only when I run `list` when I'm not cd'd into my `maildir-dir`, `~/.mail/arch`. I must've been running my tests in https://github.com/soywod/himalaya/discussions/327#discussioncomment-2313048 while in that dir.

**Reply by soywod** [reply]
> The `cannot read maildir directory "/home/user/.mail/arch/.INBOX"` error still occurs, but only when I run `list` when I'm not cd'd into my `maildir-dir`, `~/.mail/arch`. I must've been running my tests in [#327 (reply in thread)](https://github.com/soywod/himalaya/discussions/327#discussioncomment-2313048) while in that dir.

Wait I don't understand, there is no concept of "must be cd'd into my maildir". When using the maildir backend, the mailbox argument `-m|--mailbox` can be: 

- "inbox or INBOX" (case insensitive since https://github.com/soywod/himalaya/commit/d79c6c40a7e3e029cac979c6cd39cc67a5da15ba): this refers to the root of the `maildir-dir` config. According to your config, it should point to `~/.mail/arch`.
- a path (absolute, or relative to `maildir-dir` since https://github.com/soywod/himalaya/commit/1f01202262b1463b63d4678dce56779553d2e237)
- a folder name. If you give "MyFolder", then it will look for a maildir subfolder as describe in the Maildir++ spec: `~/.mail/arch/.MyFolder`

I tried again with the same config as yours and I cannot reproduce the error. I start to doubt that your `~/.mail/arch` is a real maildir, could you share the output of `ls -al ~/.mail/arch`?

**Reply by polyzen** [reply]
```
~ > ls -al ~/.mail/arch
total 32
drwxr-xr-x 8 user user 4096 Mar  9 02:28 ./
drwxr-xr-x 4 user user 4096 Mar  5 02:01 ../
drwx------ 5 user user 4096 Mar  9 02:28 Archive/
drwx------ 5 user user 4096 Mar  9 02:28 Drafts/
drwx------ 5 user user 4096 Mar  9 02:31 INBOX/
drwx------ 5 user user 4096 Mar  9 02:28 Junk/
drwx------ 5 user user 4096 Mar  9 02:28 Sent/
drwx------ 5 user user 4096 Mar  9 02:28 Trash/
```

**Reply by polyzen** [reply]
When I set `maildir-dir = "~/.mail/arch/INBOX"`, `himalaya list` works as intended.

**Reply by soywod** [reply]
It is what I thought, your `~/.mail/arch` is not a *valid* maildir directory. It is just a regular folder that contains other maildir directories. So the behaviour is totally normal. Which tool do you use to synchronize your emails locally?

**Comment by polyzen** [comment]
On March 9, 2022 4:21:31 AM EST, "Clément DOUIN" ***@***.***> wrote:
> It is what I thought, your `~/.mail/arch` is not a *valid* maildir directory. So the behaviour is totally normal. Which tool do you use to synchronize your emails locally? 
> 

I use mbsync.

--
Best,
Daniel <https://danielcapella.com>

**Reply by soywod** [reply]
I think it is related to this issue (even if it is not for the same tool): https://github.com/OfflineIMAP/offlineimap/issues/589. I think mbsync (as well as offlineimap) do not care about the Maildir++ standard. They just sync every IMAP mailboxes to their own folder. So I think the workaround is to setup aliases, sth like:

```toml
[your-account]
maildir-dir = "~/.mail/arch/INBOX"

[your-account.mailboxes]
sent = "~/.mail/arch/Sent"
drafts = "~/.mail/arch/Drafts"
# …
```

```bash
himalaya list # will show emails from the INBOX folder
himalaya -m sent list # will show emails from the Sent folder
himalaya -m drafts list # will show emails from the Drafts folder
# …

**Reply by soywod** [reply]
(I just pushed some fixes about mailbox aliases, please pull before trying again)

**Reply by soywod** [reply]
In fact it could work this way also (since paths validation is done before folder names validation):

```toml
[your-account]
maildir-dir = "~/.mail/arch"

[your-account.mailboxes]
inbox = "Inbox"
sent = "Sent"
drafts = "Drafts"
# …
```

**Reply by polyzen** [reply]
As Maildir++ is an extension of Maildir, perhaps both should be supported? mbsync optionally supports Maildir++, as does notmuch. I prefer the flat Maildir layout myself.

**Reply by polyzen** [reply]
When building the latest commit:
```
---- test_maildir_backend stdout ----
thread 'test_maildir_backend' panicked at 'called `Result::unwrap()` on an `Err` value: cannot get maildir instance from "inbox"

Caused by:
    cannot read maildir directory "/tmp/himalaya-test-mdir/.INBOX"', tests/test_maildir_backend.rs:39:51
```

**Reply by soywod** [reply]
> As Maildir++ is an extension of Maildir, perhaps both should be supported? mbsync optionally supports Maildir++, as does notmuch. I prefer the flat Maildir layout myself.

You may have misunderstood sth about Maildir.

Maildir is a standard for storing emails in a folder. There is no "link" between different Maildirs, each Maildir is independent. The fact that your `~/.mail/arch` folder contains subfolders `Inbox`, `Sent`, `Drafts` etc is just an architectural choice from `mbsync`. Btw, `~/.mail/arch` is *NOT* a valid Maildir, but it *CONTAINS* valid Maildir folders.

Maildir++ is a Maildir extension that proposes 2 major additions: folders and quotas. Maildir++ allows subfolders that start by a dot `.`. There is nothing more. A Maildir++ folder is fully compatible with a Maildir folder. Himalaya supports this extension, which means that if you put `sent = "Sent"` in your mailbox aliases, it will look for a Maildir++ subfolder `.Sent`. But if it does not find any valid folder, it will just look for a regular folder `Sent`.

No matter the tool you use to sync your emails, you will have to map folders in your Himalaya configuration (because every tool can have its own folder structure). So far this is the best approach IMO:

```toml
[your-account]
maildir-dir = "~/.mail/arch" # although, arch is not a valid maildir, but it is used as "root" maildir folder

[your-account.mailboxes]
inbox = "Inbox"
sent = "Sent"
drafts = "Drafts"
```

**Reply by soywod** [reply]
> When building the latest commit:
> 
> ```
> ---- test_maildir_backend stdout ----
> thread 'test_maildir_backend' panicked at 'called `Result::unwrap()` on an `Err` value: cannot get maildir instance from "inbox"
> 
> Caused by:
>     cannot read maildir directory "/tmp/himalaya-test-mdir/.INBOX"', tests/test_maildir_backend.rs:39:51
> ```

It looks like e2e tests broke after the recent changes, I opened an issue https://github.com/soywod/himalaya/issues/335.
